controller/rancher/certificate_fetcher.go

```diff
@@ -0,0 +1,317 @@
+	opts.Filters["removed_null"] = "1"
+
+	certs, err := fetcher.Client.Certificate.List(opts)
+	if err != nil {
+		return nil, fmt.Errorf("Coudln't get certificate by name [%s]. Error: %#v", certName, err)
+	}
+	var cert client.Certificate
+	var certWithChain string
+	if len(certs.Data) >= 1 {
+		cert = certs.Data[0]
+		certWithChain = fmt.Sprintf("%s\n%s", cert.Cert, cert.CertChain)
+	}
+	return &config.Certificate{
+		Name: cert.Name,
+		Key:  cert.Key,
+		Cert: certWithChain,
+	}, nil
+}
+
+func (fetcher *RCertificateFetcher) UpdateEndpoints(lbSvc *metadata.Service, eps []client.PublicEndpoint) error {
+	opts := client.NewListOpts()
+	opts.Filters["uuid"] = lbSvc.UUID
+	opts.Filters["removed_null"] = "1"
+	lbs, err := fetcher.Client.LoadBalancerService.List(opts)
+	if err != nil {
+		return fmt.Errorf("Coudln't get LB service by uuid [%s]. Error: %#v", lbSvc.UUID, err)
+	}
+	if len(lbs.Data) == 0 {
+		logrus.Infof("Failed to find lb by uuid %s", lbSvc.UUID)
+		return nil
+	}
+	lb := lbs.Data[0]
+
+	toUpdate := make(map[string]interface{})
+	toUpdate["publicEndpoints"] = eps
+	logrus.Infof("Updating Rancher LB [%s] in stack [%s] with the new public endpoints [%v] ", lbSvc.Name, lbSvc.StackName, eps)
+	if _, err := fetcher.Client.LoadBalancerService.Update(&lb, toUpdate); err != nil {
+		return fmt.Errorf("Failed to update Rancher LB [%s] in stack [%s]. Error: %#v", lbSvc.Name, lbSvc.StackName, err)
+	}
+	return nil
+}
+
+/* Cache the certs present under CertDir and DefaultCertDir
+   When the Cache is null, read from the dir and init the cache. When cache is not empty read from the cache.
+   Clear cache when the directories change
+   Start a Watcher thread when initialising the cache. For every certName, add to watcher.
+   When watcher receives event, read that cert and update cache.
+   Add a watcher for CertDir/DefaultCertDir also - upon events on these dirs clear the cache and enque.*/
+
+func (fetcher *RCertificateFetcher) ReadAllCertificatesFromDir(certDir string) ([]*config.Certificate, error) {
+	if _, err := os.Stat(certDir); os.IsNotExist(err) {
+		return nil, fmt.Errorf("Certificate Directory does not exist %v", certDir)
+	}
+	certs := []*config.Certificate{}
+
+	fetcher.mu.RLock()
+	currentCertD := fetcher.CertDir
+	fetcher.mu.RUnlock()
+
+	if currentCertD != "" && currentCertD == certDir && fetcher.CertsCache != nil && len(fetcher.CertsCache) > 0 {
+		//read from cache if certDir matches and certs are cached.
+		logrus.Debug("Reading certs from cache")
+	} else {
+		// init/re-init cache
+		fetcher.mu.Lock()
+		fetcher.removeWatches()
+
+		fetcher.CertsCache = make(map[string]*config.Certificate)
+		logrus.Debug("Reading certs from directory and starting watchers")
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(certDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.CertDir = ""
+			fetcher.CertsCache = nil
+			logrus.Errorf("Error %v reading certs from directory %v", err, certDir)
```

Comment:
@prachidamle we better error out and NOT reload the config which is already running with the previously fetched certificates.